explorer/server.go

```diff
@@ -0,0 +1,210 @@
+package explorer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/zos/client"
+	"github.com/threefoldtech/zos/pkg/rmb"
+)
+
+func (a *App) listFarms(w http.ResponseWriter, r *http.Request) {
+	r, err := a.handleRequestsQueryParams(r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+		return
+	}
+	maxResult, pageOffset := getMaxResult(r.Context()), getOffset(r.Context())
+
+	queryString := fmt.Sprintf(`
+	{
+		farms (limit:%d,offset:%d) {
+			name
+			farmId
+			twinId
+			version
+			farmId
+			pricingPolicyId
+		}
+		publicIps{
+			id
+			ip
+			farmId
+			contractId
+			gateway
+			
+		}
+	}
+	`, maxResult, pageOffset)
+
+	_, err = queryProxy(queryString, a.explorer, w)
+
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+	}
+}
+
+func (a *App) listNodes(w http.ResponseWriter, r *http.Request) {
+	r, err := a.handleRequestsQueryParams(r)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+	}
+
+	maxResult := getMaxResult(r.Context())
+	pageOffset := getOffset(r.Context())
+	isSpecificFarm := getSpecificFarm(r.Context())
+
+	queryString := fmt.Sprintf(`
+	{
+		nodes(limit:%d,offset:%d,%s){
+			version          
+			id
+			nodeId        
+			farmId          
+			twinId          
+			country
+			gridVersion  
+			city         
+			uptime           
+			created          
+			farmingPolicyId
+			updatedAt
+			cru
+			mru
+			sru
+			hru
+		publicConfig{
+			gw4
+			ipv4
+			ipv6
+			gw6
+		  }
+		}
+	}
+	`, maxResult, pageOffset, isSpecificFarm)
+
+	_, err = queryProxy(queryString, a.explorer, w)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+	}
+}
+
+func (a *App) getNode(w http.ResponseWriter, r *http.Request) {
+
+	nodeID := mux.Vars(r)["node_id"]
+	value, _ := a.GetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
+
+	// No value, fetch data from the node
+	if value == "" {
+		nodeInfo, err := a.fetchNodeData(r.Context(), nodeID)
+		if err != nil {
+			log.Error().Err(err).Msg("could not fetch node data")
+			w.WriteHeader(http.StatusBadRequest)
```

Comment:
should be `internal server error`?

Reply:
It's definitely not a BadRequest, but Why not `NotFound` ? the truth is the fetchNodeData should more specific about the error it returned 
for example, if it failed to connect to the graphQL api, then that is an internal server error. But if it managed to make the query but the result where empty. it's then a NotFound.

So how to do this?
With error wrapping.

for example
```go
package main

import (
	"errors"
	"fmt"
)

var (
	ErrNodeNotFound = errors.New("node not found")
)

func doSomething() error {
	if nodeNotFound {
		return ErrNodeNotFound
	}

	if otherUnknwonError {
		return fmt.Errorf("something else happend")
	}
	// all is good 
	return nil
}

func main() {
	err := doSomething()

	if errors.Is(err, ErrNodeNotFound) {
		// return not found 404
	} else if err != nil {
		// return internal server error
	}

	// err is nil, and al is good
}

```

it's recommended you read about the `errors` package in Go stdlibrary . it also explain how u can wrap errors and still detect the underlying cause with `errors.Is`